File: dotnet/src/Frontend/src/widgets/NavBar.js
import * as React from "react";
import SkyLight from 'react-skylight';
import * as util from "../Util";
import CreateProject from '../modals/CreateProject';

let navBarInstance = null;

export function showModal(content, props) {
  if (navBarInstance == null) {
      throw new Error("No NavBar instance available");
  }
  return navBarInstance.showModal(content, props);
}

export default class NavBar extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
        modal: null,
        isBurgerOpen: false,
        serviceInfo: props.global.state.serviceInfo,
        useRightClick: props.global.state.useRightClick
    };
  }

  componentDidMount() {
    navBarInstance = this;
    this.disposable =
        this.props.global.subscribe("serviceInfo", serviceInfo =>
            this.setState({serviceInfo}));
  }

  componentWillUnmount() {
    navBarInstance = null;
    if (this.disposable) {
      this.disposable.dispose();
    }
  }

  onClick(dropdownIndex) {
    return () => {
      switch (dropdownIndex) {
          case 0:
            showModal(CreateProject);
            break;
          case 1:
            util.loadProject();
            break;
          case 2:
            IrisLib.saveProject();
            break;
          case 3:
            IrisLib.unloadProject();
            break;
          case 4:
            IrisLib.shutdown();
            break;
          case 5:
            var useRightClick = !this.state.useRightClick;
            this.props.global.useRightClick(useRightClick);
            this.setState({useRightClick});
            break;
      }
    }
  }

  showModal(content, props) {
    return new Promise(resolve =>
        this.setState({
            modal: {
                content,
                props: props || {},
                onSubmit: resolve
            }
        }, () => this.modalEl.show())
    );
  }

  renderModal() {
    if (this.state.modal == null) {
        return null;
    }

    return React.createElement(this.state.modal.content, {
      onSubmit: values => {
        this.modalEl.hide();
        this.state.modal.onSubmit(values);
      },
      ...this.state.modal.props
    });
  }

  render() {
    return (
      <div>
        <SkyLight
            hideOnOverlayClicked
            ref={el => this.modalEl = el}
            title={this.state.modal ? this.state.modal.title : ""}
            dialogStyles={{height: "inherit"}}>
            {this.renderModal()}
        </SkyLight>
        <nav id="app-header" className="navbar ">
          <div className="navbar-brand">
            <a className="navbar-item" href="http://nsynk.de">
              <img src="lib/img/nsynk.png" />
            </a>

             <div className="navbar-burger burger"
                onClick={_ => this.setState({isBurgerOpen: !this.state.isBurgerOpen})} >
              <span></span>
              <span></span>
              <span></span>
            </div>
          </div>

          <div className={"navbar-menu " + (this.state.isBurgerOpen ? "is-active" : "")}>
            <div className="navbar-start">
              <div className="navbar-item has-dropdown is-hoverable">
                <a className="navbar-link" style={{fontSize: "14px"}}>Iris Menu</a>
                <div className="navbar-dropdown ">
                  <a className="navbar-item" onClick={this.onClick(0).bind(this)}>Create Project</a>
                  <a className="navbar-item" onClick={this.onClick(1).bind(this)}>Load Project</a>
                  <a className="navbar-item" onClick={this.onClick(2).bind(this)}>Save Project</a>
                  <a className="navbar-item" onClick={this.onClick(3).bind(this)}>Unload Project</a>
                  <a className="navbar-item" onClick={this.onClick(4).bind(this)}>Shutdown</a>
                  <a className="navbar-item" onClick={this.onClick(5).bind(this)}>{"Use right click: " + this.state.useRightClick}</a>
                </div>
              </div>
            </div>
            <div className="navbar-end">
              <div className="navbar-item">
                Iris v{this.state.serviceInfo.version} - build {this.state.serviceInfo.buildNumber}
              </div>
            </div>
          </div>
        </nav>
      </div>
    )
  }
}

File: dotnet/src/Frontend/fable/Frontend/Elmish/Navbar.fs
module Iris.Web.Navbar

open Fable.Helpers.React
open Fable.Import
open Fable.Core
open Fable.Core.JsInterop
open Fable.PowerPack
open Iris.Core
open Helpers
open Types

type IProjectInfo =
  abstract name: Name
  abstract username: UserName
  abstract password: Password

let CreateProjectModal: React.ComponentClass<ModalProps<obj, string>> =
  importDefault "../../../src/modals/CreateProject"

let LoadProjectModal: React.ComponentClass<ModalProps<obj, IProjectInfo>> =
  importDefault "../../../src/modals/LoadProject"

let ProjectConfigModal: React.ComponentClass<ModalProps<string[], string>> =
  importDefault "../../../src/modals/ProjectConfig"

module Options =
  let [<Literal>] createProject = "Create Project"
  let [<Literal>] loadProject = "Load Project"
  let [<Literal>] saveProject = "Save Project"
  let [<Literal>] unloadProject = "Unload Project"
  let [<Literal>] shutdown = "Shutdown"

let onClick dispatch id _ =
  let start f msg =
    f() |> Promise.iter (fun () -> printfn "%s" msg)
  match id with
  | Options.createProject ->
    makeModal dispatch CreateProjectModal None
    |> Promise.iter Lib.createProject
  | Options.loadProject ->
    promise {
      let! info = makeModal dispatch LoadProjectModal None
      let! err = Lib.loadProject(info.name, info.username, info.password, None, None)
      match err with
      | Some err ->
        // Get project sites and machine config
        let! sites = Lib.getProjectSites(info.name, info.username, info.password)
        // Ask user to create or select a new config
        let! site = makeModal dispatch ProjectConfigModal (Some sites)
        // Try loading the project again with the site config
        let! err2 = Lib.loadProject(info.name, info.username, info.password, Some (Id site), None)
        err2 |> Option.iter (printfn "Error when loading site %s: %s" site)
      | None -> ()
    } |> Promise.start
  | Options.saveProject ->
    start Lib.saveProject "Project has been saved"
  | Options.unloadProject ->
    start Lib.unloadProject "Project has been unloaded"
  | Options.shutdown ->
    start Lib.shutdown "Iris has been shut down"
  | other ->
    failwithf "Unknow navbar option: %s" other

open Fable.Helpers.React.Props

let dropdown dispatch (model: Model) =
  let navbarItem opt =
    a [Class "navbar-item"; OnClick (onClick dispatch opt)] [str opt]
  div [Class "navbar-item has-dropdown is-hoverable"] [
    a [
      Class "navbar-link"
      Style [!!("fontSize", "14px")]
    ] [str "Iris Menu"]
    div [Class "navbar-dropdown"] [
      navbarItem Options.createProject
      navbarItem Options.loadProject
      navbarItem Options.saveProject
      navbarItem Options.unloadProject
      navbarItem Options.shutdown
      a [
        Class "navbar-item"
        OnClick (fun _ ->
          { model.userConfig with useRightClick = not model.userConfig.useRightClick }
          |> UpdateUserConfig |> dispatch)
      ] [str ("Use right click: " + (string model.userConfig.useRightClick))]
    ]
  ]

let view dispatch (model: Model) =
  let version, buildNumber =
    match model.state with
    | Some state ->
      try
        let info = Iris.Web.Core.Client.ClientContext.Singleton.ServiceInfo
        info.version, info.buildNumber
      with ex ->
        // printfn "Cannot read ServiceInfo from ClientContext"
        "0.0.0", "123"
    | None -> "0.0.0", "123"
  div [] [
    nav [Id "app-header"; Class "navbar "] [
      div [Class "navbar-brand"] [
        a [Class "navbar-item"; Href "http://nsynk.de"] [
          img [Src "lib/img/nsynk.png"]
        ]
      ]
      div [Class "navbar-menu is-active"] [
        div [Class "navbar-start"] [
          dropdown dispatch model
        ]
        div [Class "navbar-end"] [
          div [Class "navbar-item"] [
            str(sprintf "Iris v%s - build %s" version buildNumber)
          ]
        ]
      ]
    ]
  ]
